lib/shard_reader.rb

```diff
@@ -0,0 +1,85 @@
+class ShardReader
+
+  def initialize(stream_name:,
+                 tracker_prefix:,
+                 shard_id:,
+                 batch_size:,
+                 logger:,
+                 client:)
+    @stream_name = stream_name
+    @tracker_prefix = tracker_prefix
+    @shard_id = shard_id
+    @batch_size = batch_size
+    @logger = logger
+    @client = client
+    @stop_processing = false
+
+    key_prefix = [ @stream_name,
+                   @tracker_prefix,
+                   shard_id ].compact.join('-')
+    @tracker = SequenceNumberTracker.new(key_prefix: key_prefix)
+  end
+
+  def run(&block)
+    @thread = Thread.new do
+      loop do
+        break if @stop_processing
+        begin
+          iterator_opts = { stream_name: @stream_name,
+                            shard_id: @shard_id }
+          if seq = @tracker.last_sequence_number
+            iterator_opts[:shard_iterator_type] = 'AFTER_SEQUENCE_NUMBER'
+            iterator_opts[:starting_sequence_number] = seq
+          else
+            iterator_opts[:shard_iterator_type] = 'TRIM_HORIZON'
+          end
+          @logger.debug "Getting shard iterator for #{@stream_name} / #{@shard_id} / #{seq}"
+
+          resp = @client.get_shard_iterator(iterator_opts)
+          shard_iterator = resp.shard_iterator
+
+          # Iterate!
+          loop do
+            break if @stop_processing
+            sleep 1
+            @logger.debug "Getting records for #{shard_iterator}"
+            resp = @client.get_records(shard_iterator: shard_iterator,
+                                       limit: @batch_size)
+
+            resp.records.each do |record|
+              process_record(record, resp, &block)
+            end
+            shard_iterator = resp.next_shard_iterator
```

Comment:
It's a little funky - the outer loop sets up the initial shard iterator based on whether it has a sequence number to resume from or not. From there, we can use that shard iterator cursor to get our first batch of records. The `get_records` request always gives you back the next shard iterator cursor as well, so you have a new place to make your next request from. The inner loop structure just keeps it going, getting the next set of records as long as we can grab it. The assignment basically just sets up the next iteration's `get_records` call.

The outer loop also helps with error handling - that way, if we get into a spot where the iterator we're pushing forward on each `get_records` call is no longer valid, we can reset and get a new one based on our saved sequence number.